# HW2

### due on 10/8/2017 at 11:59 pm

<b>Instructions:</b> Write your answers in the empty code cells below.

<b>Penalties:</b> You will incur penalties if:
<ul>
<li>Your code is wrong</li>
<li>Your code does not work on different data</li>
<li>You use for loops (unless explicitly permitted)</li>
</ul>

## Problem

Start by executing the following code, which creates a Series object <b>stock</b> of share prices of Nvidia. 

### If the following code works, then you are good...

In [1]:
import pandas as pd
import fix_yahoo_finance as yf
data = yf.download("NVDA", start="1999-01-01", end="2017-10-01")
stock = data.Close

/Users/Meiyuan/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/fix_yahoo_finance/__init__.py:43: DeprecationWarning: 
    Auto-overriding of pandas_datareader's get_data_yahoo() is deprecated and no longer available.
    Use pdr_override() to explicitly override it.
  DeprecationWarning)


[*********************100%***********************]  1 of 1 downloaded

### Question 1


What was the closing price on Jul 2, 2012?

In [2]:
stock['2012-07-02']

/Users/Meiyuan/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/IPython/core/formatters.py:92: DeprecationWarning: DisplayFormatter._ipython_display_formatter_default is deprecated: use @default decorator instead.
  def _ipython_display_formatter_default(self):
/Users/Meiyuan/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/IPython/core/formatters.py:98: DeprecationWarning: DisplayFormatter._formatters_default is deprecated: use @default decorator instead.
  def _formatters_default(self):
/Users/Meiyuan/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/IPython/core/formatters.py:677: DeprecationWarning: PlainTextFormatter._deferred_printers_default is deprecated: use @default decorator instead.
  def _deferred_printers_default(self):
/Users/Meiyuan/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/IPython/core/formatters.py:669: DeprecationWarning: PlainTextFormatter._singleton_printers_default is deprecated: use @default decora

13.449999999999999

### Question 2

Find the ten largest <b>one-day % drops</b> (in % compared to the closing cost of the previous trading day). Make sure that your output is a Series object that shows both date and the % drop. 

In [3]:
stockShiftOne = stock.shift(-1) 

In [4]:
oneDayDiff = stockShiftOne - stock

In [5]:
(oneDayDiff / stock).nsmallest(10) * 100

Date
2004-08-05   -35.233519
2002-07-30   -31.750929
2008-07-02   -30.726571
2000-03-13   -26.214408
2002-11-07   -21.924392
2003-08-07   -19.689110
2008-02-13   -16.321240
2000-11-29   -16.062180
2002-06-18   -15.664775
2001-01-04   -14.971462
Name: Close, dtype: float64

### Question 3

In which days did we observe that the price has increased of at least 60% within the 9 preceding trading sessions? Report your result as a Series (d,gain) where <i>d</i> is the trading session date when the increase was observed and <i>gain</i> is the maximum percentage gain (%) observed over the 9 days preceding session <i>d</i>. <i>Hint:</i> use a rolling minimum

Creat a 9-day rolling window and find out the minimum value within every 9-day session

In [6]:
prev9Days = stock.rolling(9)
nineDaysMin = prev9Days.min()

Computet the profit/loss

In [7]:
profit = ((stock - nineDaysMin) / nineDaysMin)

Inspect the data

In [8]:
profit.nlargest(15)

Date
2000-03-13    1.576052
2000-03-10    1.036677
2000-03-14    0.900755
2000-03-15    0.805555
2000-03-09    0.731121
2001-10-12    0.693048
2000-02-18    0.689655
2000-03-16    0.609774
2001-01-11    0.603340
2000-02-22    0.602446
2001-10-11    0.589615
2000-02-17    0.586207
2001-01-12    0.569937
2001-01-04    0.565022
2002-11-21    0.540076
Name: Close, dtype: float64

boolean mask to filter out those with at least 60% increase

In [9]:
mask = profit > 0.6

Use mask to display the result

In [10]:
profit[mask] * 100

Date
2000-02-18     68.965490
2000-02-22     60.244639
2000-03-09     73.112067
2000-03-10    103.667739
2000-03-13    157.605178
2000-03-14     90.075506
2000-03-15     80.555549
2000-03-16     60.977436
2001-01-11     60.334040
2001-10-12     69.304849
Name: Close, dtype: float64

### Question 4

What was the worst possible "buy&sell" operation? Find the two days x and y so that the % loss resulting from buying in x and selling in y (with y after x) is maximized. Your code should work for any data. <i>Hint:</i> the method <b>cummin</b> returns the minimum value encountered so far. Make sure to return the trading days x and y, as well as the % loss resulting from buying in x and selling in y.

Construct a new Series to store the highest buying price up till the last day except today


In [11]:
cumMax = stock.shift(1).cummax()

In [12]:
cumMax[:5] # highest buy price up till today, 1.640625 on 1999-01-25 means the highest buy price before 1999-01-25 is 1.640625 

Date
1999-01-22         NaN
1999-01-25    1.640625
1999-01-26    1.812500
1999-01-27    1.812500
1999-01-28    1.812500
Name: Close, dtype: float64

In [13]:
stock[:5] # sell price if stock is sold on that day

Date
1999-01-22    1.640625
1999-01-25    1.812500
1999-01-26    1.671875
1999-01-27    1.666667
1999-01-28    1.661458
Name: Close, dtype: float64

profit = (Sell price -  buy price)/ buy price

In [14]:
((stock - cumMax)/cumMax).nsmallest(1) # get the smallest profit and sell happen date

Date
2002-10-09   -0.897225
Name: Close, dtype: float64

In [15]:
stock['2002-10-09'] # get the sell price

2.4566669999999999

In [16]:
cumMax['2002-10-09'] # get the buy price

23.903334000000001

To get both of the dates, we can easily get Sell time stamp first

In [17]:
sellTime = ((stock - cumMax)/cumMax).idxmin()

In [18]:
sellTime

Timestamp('2002-10-09 00:00:00')

Use a boolean mask to fitler all date before the sell happened and pick the biggest one

In [19]:
stock [stock.index < sellTime].idxmax()

Timestamp('2002-01-03 00:00:00')

## Observation
### "The worst buy-sell operation is if you buy on 2002-01-03(23.903334000000001) and sell on 2002-10-08(2.4566669999999999), which causes a loss of 89%"